File: mulle/rules/scoring.py
from ..models.player import Player

INTAKE_POINTS_1 = {"SP": list(map(str, range(3,14))) + ["A"], "RU": ["A"], "HJ": ["A"], "KL": ["A"]}
INTAKE_POINTS_2 = {"SP": ["2", "A"], "RU": ["10"],}

def intake_points(player: Player) -> int:
    pts = 0
    for c in player.captured:
        if c.suit in INTAKE_POINTS_1 and c.rank in INTAKE_POINTS_1[c.suit]:
            pts += 1
        if c.suit in INTAKE_POINTS_2 and c.rank in INTAKE_POINTS_2[c.suit]:
            pts += 2
    return pts

class ScoreBreakdown:
    def __init__(self, player: Player, mulle_points: int, tabbe: int, intake: int, bonus: int, total: int):
        self.player = player
        self.mulle_points = mulle_points
        self.tabbe = tabbe
        self.intake = intake
        self.bonus = bonus
        self.total = total

    def __repr__(self):
        return f"Score({self.player.name}: mulle={self.mulle_points}, tabbe={self.tabbe}, intake={self.intake}, bonus={self.bonus}, total={self.total})"


def score_round(players: list[Player]) -> list[ScoreBreakdown]:
    breakdowns: list[ScoreBreakdown] = []
    for p in players:
        mulle_pts = p.total_mulle_points()
        tabbe_pts = p.tabbe
        intake_pts = intake_points(p)
        bonus = 0
        if intake_pts > 20:
            bonus = (intake_pts - 20) * 2
        total = mulle_pts + tabbe_pts + intake_pts + bonus
        breakdowns.append(ScoreBreakdown(p, mulle_pts, tabbe_pts, intake_pts, bonus, total))
    return breakdowns

File: src/rules/scoring.ts
import { Player } from '../models/Player';

/**
 * Cards that give 1 intake point
 */
export const INTAKE_POINTS_1: Record<string, string[]> = {
  'SP': ['3', '4', '5', '6', '7', '8', '9', '10', 'J', 'Q', 'K'],
  'RU': ['A'],
  'HJ': ['A'],
  'KL': ['A']
};

/**
 * Cards that give 2 intake points
 */
export const INTAKE_POINTS_2: Record<string, string[]> = {
  'SP': ['2', 'A'],
  'RU': ['10']
};

/**
 * Calculate intake points for a player.
 * Intake points are used for bonus calculation but NOT included in total score.
 */
export function intakePoints(player: Player): number {
  let pts = 0;
  
  for (const card of player.captured) {
    if (card.suit in INTAKE_POINTS_1 && INTAKE_POINTS_1[card.suit].includes(card.rank)) {
      pts += 1;
    }
    if (card.suit in INTAKE_POINTS_2 && INTAKE_POINTS_2[card.suit].includes(card.rank)) {
      pts += 2;
    }
  }
  
  return pts;
}

/**
 * Score breakdown for a player
 */
export class ScoreBreakdown {
  player: Player;
  mullePoints: number;
  tabbe: number;
  intake: number;
  bonus: number;
  total: number;

  constructor(
    player: Player,
    mullePoints: number,
    tabbe: number,
    intake: number,
    bonus: number,
    total: number
  ) {
    this.player = player;
    this.mullePoints = mullePoints;
    this.tabbe = tabbe;
    this.intake = intake;
    this.bonus = bonus;
    this.total = total;
  }

  toString(): string {
    return `Score(${this.player.name}: mulle=${this.mullePoints}, tabbe=${this.tabbe}, intake=${this.intake}, bonus=${this.bonus}, total=${this.total})`;
  }
}

/**
 * Score a round for all players.
 * 
 * Scoring:
 * - Mulle: Sum of all mulle pair values (A=14, J=11, Q=12, K=13, numbers=face value)
 * - Tabbe: Number of captures from empty board
 * - Intake: Sum of intake points (used for bonus, NOT in total)
 * - Bonus: (intake - 20) × 2 if intake > 20
 * - TOTAL: mulle + tabbe + bonus (intake NOT included)
 */
export function scoreRound(players: Player[]): ScoreBreakdown[] {
  const breakdowns: ScoreBreakdown[] = [];
  
  for (const p of players) {
    const mullePts = p.totalMullePoints();
    const tabbePts = p.tabbe;
    const intakePts = intakePoints(p);
    
    let bonus = 0;
    if (intakePts > 20) {
      bonus = (intakePts - 20) * 2;
    }
    
    // Total = mulle + tabbe + bonus (intake NOT counted in total)
    const total = mullePts + tabbePts + bonus;
    
    breakdowns.push(new ScoreBreakdown(p, mullePts, tabbePts, intakePts, bonus, total));
  }
  
  return breakdowns;
}
